File: Backend/paste/handlers.go
package paste

import (
	"encoding/json"
	"errors"
	"net/http"
	"time"

	"golang.org/x/crypto/bcrypt"
)

type Handlers struct {
	store *Store
}

func NewHandlers(store *Store) *Handlers {
	return &Handlers{store: store}
}

type createRequest struct {
	Title    string `json:"title"`
	Content  string `json:"content"`
	Language string `json:"language"`
	Expiry   int    `json:"expiry"`
	Password string `json:"password"`
}

func writeJSON(w http.ResponseWriter, status int, v any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(v)
}

func writeError(w http.ResponseWriter, status int, msg string) {
	writeJSON(w, status, map[string]string{"error": msg})
}

func (h *Handlers) Create(w http.ResponseWriter, r *http.Request) {
	r.Body = http.MaxBytesReader(w, r.Body, 1_100_000)

	var req createRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	if req.Content == "" {
		writeError(w, http.StatusBadRequest, "content is required")
		return
	}
	if len(req.Content) > 1_000_000 {
		writeError(w, http.StatusRequestEntityTooLarge, "content exceeds 1,000,000 characters")
		return
	}
	if len(req.Title) > 200 {
		writeError(w, http.StatusBadRequest, "title exceeds 200 characters")
		return
	}

	var passwordHash string
	if req.Password != "" {
		if len(req.Password) > 72 {
			writeError(w, http.StatusBadRequest, "password exceeds 72 characters")
			return
		}
		hash, err := bcrypt.GenerateFromPassword([]byte(req.Password), bcrypt.DefaultCost)
		if err != nil {
			writeError(w, http.StatusInternalServerError, "failed to process password")
			return
		}
		passwordHash = string(hash)
	}

	p := &Paste{
		Title:        req.Title,
		Content:      req.Content,
		Language:     req.Language,
		PasswordHash: passwordHash,
		Protected:    passwordHash != "",
	}

	if req.Expiry < 1 || req.Expiry > 10 {
		writeError(w, http.StatusBadRequest, "expiry must be between 1 and 10 minutes")
		return
	}
	t := time.Now().Add(time.Duration(req.Expiry) * time.Minute)
	p.ExpiresAt = &t

	if err := h.store.Create(p); err != nil {
		writeError(w, http.StatusInternalServerError, "failed to create paste")
		return
	}

	writeJSON(w, http.StatusCreated, map[string]string{
		"id":  p.ID,
		"url": "/paste/" + p.ID,
	})
}

func (h *Handlers) Get(w http.ResponseWriter, r *http.Request) {
	id := r.PathValue("id")
	p, err := h.store.Get(id)
	if errors.Is(err, ErrNotFound) {
		writeError(w, http.StatusNotFound, "paste not found")
		return
	}
	if err != nil {
		writeError(w, http.StatusInternalServerError, "failed to retrieve paste")
		return
	}
	if p.ExpiresAt != nil && time.Now().After(*p.ExpiresAt) {
		h.store.Delete(id)
		writeError(w, http.StatusGone, "paste has expired")
		return
	}
	if p.Protected {
		pw := r.Header.Get("X-Paste-Password")
		if pw == "" || bcrypt.CompareHashAndPassword([]byte(p.PasswordHash), []byte(pw)) != nil {
			writeJSON(w, http.StatusForbidden, map[string]any{"protected": true, "error": "password required"})
			return
		}
	}
	writeJSON(w, http.StatusOK, p)
}

func (h *Handlers) GetRaw(w http.ResponseWriter, r *http.Request) {
	id := r.PathValue("id")
	p, err := h.store.Get(id)
	if errors.Is(err, ErrNotFound) {
		http.Error(w, "paste not found", http.StatusNotFound)
		return
	}
	if err != nil {
		http.Error(w, "failed to retrieve paste", http.StatusInternalServerError)
		return
	}
	if p.ExpiresAt != nil && time.Now().After(*p.ExpiresAt) {
		h.store.Delete(id)
		http.Error(w, "paste has expired", http.StatusGone)
		return
	}
	if p.Protected {
		pw := r.Header.Get("X-Paste-Password")
		if pw == "" || bcrypt.CompareHashAndPassword([]byte(p.PasswordHash), []byte(pw)) != nil {
			http.Error(w, "password required", http.StatusForbidden)
			return
		}
	}
	w.Header().Set("Content-Type", "text/plain; charset=utf-8")
	w.Header().Set("X-Content-Type-Options", "nosniff")
	w.Write([]byte(p.Content))
}

File: backend/src/paste/handlers.rs
use axum::{
    Json,
    extract::{Path, State},
    http::{HeaderMap, StatusCode},
    response::IntoResponse,
};
use chrono::{Duration, Utc};
use serde::Deserialize;
use serde_json::{Value, json};
use uuid::Uuid;

use crate::{
    api_error::ApiError,
    paste::store::{Paste, Store},
};

#[derive(Deserialize)]
pub struct CreateRequest {
    title: String,
    content: String,
    language: String,
    expiry: i64,
    password: Option<String>,
}

pub(crate) async fn create_paste(
    State(store): State<Store>,
    Json(req): Json<CreateRequest>,
) -> Result<impl IntoResponse, ApiError> {
    if req.content.is_empty() {
        return Err(ApiError::BadRequest("content is required".into()));
    }
    if req.content.len() > 1_000_000 {
        return Err(ApiError::BadRequest(
            "content exceeds 1,000,000 characters".into(),
        ));
    }
    if req.title.len() > 200 {
        return Err(ApiError::BadRequest("title exceeds 200 characters".into()));
    }
    if req.expiry < 1 || req.expiry > 10 {
        return Err(ApiError::BadRequest(
            "expiry must be between 1 and 10 minutes".into(),
        ));
    }

    let (password_hash, protected) = if let Some(pw) = req.password {
        if pw.len() > 72 {
            return Err(ApiError::BadRequest(
                "password exceeds 72 characters".into(),
            ));
        }
        let hash = bcrypt::hash(pw, bcrypt::DEFAULT_COST).map_err(|_| ApiError::InternalError)?;
        (hash, true)
    } else {
        (String::new(), false)
    };

    let paste = Paste {
        id: String::new(),
        title: req.title,
        content: req.content,
        language: req.language,
        created_at: Utc::now(),
        expires_at: Some(Utc::now() + Duration::minutes(req.expiry)),
        password_hash,
        protected,
    };

    let paste = store.create(paste).await?;

    Ok((
        StatusCode::CREATED,
        Json(json!({
            "id": paste.id,
            "url": format!("/paste/{}", paste.id),
        })),
    ))
}

pub(crate) async fn get_paste(
    State(store): State<Store>,
    Path(id): Path<String>,
    headers: HeaderMap,
) -> Result<Json<Value>, ApiError> {
    let id = Uuid::parse_str(&id).map_err(|_| ApiError::BadRequest("invalid id".into()))?;
    let paste = store.get(id).await?;

    if paste.expires_at.map_or(false, |exp| Utc::now() > exp) {
        store.delete(id).await.ok();
        return Err(ApiError::Gone);
    }

    if paste.protected {
        let pw = headers
            .get("X-Paste-Password")
            .and_then(|v| v.to_str().ok())
            .unwrap_or("");
        if bcrypt::verify(pw, &paste.password_hash).unwrap_or(false) == false {
            return Err(ApiError::Forbidden);
        }
    }

    Ok(Json(
        serde_json::to_value(&paste).map_err(|_| ApiError::InternalError)?,
    ))
}

pub(crate) async fn get_raw_paste(
    State(store): State<Store>,
    Path(id): Path<String>,
    headers: HeaderMap,
) -> Result<impl IntoResponse, ApiError> {
    let id = Uuid::parse_str(&id).map_err(|_| ApiError::BadRequest("invalid id".into()))?;
    let paste = store.get(id).await?;

    if paste.expires_at.map_or(false, |exp| Utc::now() > exp) {
        store.delete(id).await.ok();
        return Err(ApiError::Gone);
    }

    if paste.protected {
        let pw = headers
            .get("X-Paste-Password")
            .and_then(|v| v.to_str().ok())
            .unwrap_or("");
        if bcrypt::verify(pw, &paste.password_hash).unwrap_or(false) == false {
            return Err(ApiError::Forbidden);
        }
    }

    Ok((
        StatusCode::OK,
        [
            ("Content-Type", "text/plain; charset=utf-8"),
            ("X-Content-Type-Options", "nosniff"),
        ],
        paste.content,
    ))
}
